crates/nu-path/src/pwd_per_drive.rs

```diff
@@ -0,0 +1,349 @@
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Debug)]
+pub enum PathError {
+    InvalidDriveLetter,
+    InvalidPath,
+    CantLockSharedMap,
+}
+/// Usage for pwd_per_drive on windows
+///
+/// Upon change PWD, call set_pwd() with absolute path
+///
+/// Call expand_pwd() with relative path to get absolution path
+///
+/// ```
+/// use std::path::{Path, PathBuf};
+/// use nu_path::{expand_pwd, set_pwd};
+///
+/// // Set PWD for drive C
+/// set_pwd(Path::new(r"C:\Users\Home")).unwrap();
+///
+/// // Expand a relative path
+/// let expanded = expand_pwd(Path::new("c:test"));
+/// assert_eq!(expanded, Some(PathBuf::from(r"C:\Users\Home\test")));
+///
+/// // Will NOT expand an absolute path
+/// let expanded = expand_pwd(Path::new(r"C:\absolute\path"));
+/// assert_eq!(expanded, None);
+///
+/// // Expand with no drive letter
+/// let expanded = expand_pwd(Path::new(r"\no_drive"));
+/// assert_eq!(expanded, None);
+///
+/// // Expand with no PWD set for the drive
+/// let expanded = expand_pwd(Path::new("D:test"));
+/// assert!(expanded.is_some());
+/// let abs_path = expanded.unwrap().as_path().to_str().expect("OK").to_string();
+/// assert!(abs_path.starts_with(r"D:\"));
+/// assert!(abs_path.ends_with(r"\test"));
+/// ```
+pub mod singleton {
```

Comment:
can we rename this something other than singleton please?

Reply:
is 'global' better?

Reply:
not really. maybe pwd_per_drive_<something>. not sure what the <something> should be though

Reply:
Try 'shared_map'

Reply:
Rename 'singleton' to 'shared'

Reply:
I'm good with shared_map or pwd_shared_map.

Reply:
OK, renamed to 'shared_map'